dhctl/pkg/system/ssh/frontend/reverse-tunnel.go

```diff
@@ -16,68 +16,227 @@ package frontend
 
 import (
 	"fmt"
+	"math/rand"
+	"os"
 	"os/exec"
-	"strings"
+	"sync"
+	"time"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/util/retry"
 
 	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
 	"github.com/deckhouse/deckhouse/dhctl/pkg/system/ssh/cmd"
 	"github.com/deckhouse/deckhouse/dhctl/pkg/system/ssh/session"
 )
 
+type ReverseTunnelChecker interface {
+	CheckTunnel() (string, error)
+}
+
+type ReverseTunnelKiller interface {
+	KillTunnel() (string, error)
+}
+
+type tunnelWaitResult struct {
+	id  int
+	err error
+}
+
 type ReverseTunnel struct {
 	Session *session.Session
 	Address string
-	sshCmd  *exec.Cmd
+
+	tunMutex sync.Mutex
+	sshCmd   *exec.Cmd
+	started  bool
+	stopCh   chan struct{}
+
+	errorCh chan tunnelWaitResult
 }
 
 func NewReverseTunnel(sess *session.Session, address string) *ReverseTunnel {
 	return &ReverseTunnel{
 		Session: sess,
 		Address: address,
+		errorCh: make(chan tunnelWaitResult),
 	}
 }
 
 func (t *ReverseTunnel) Up() error {
-	if t.Session == nil {
-		return fmt.Errorf("up tunnel '%s': SSH client is undefined", t.String())
+	_, err := t.upNewTunnel(-1)
+	return err
+}
+
+func (t *ReverseTunnel) upNewTunnel(oldId int) (int, error) {
+	t.tunMutex.Lock()
+	defer t.tunMutex.Unlock()
+
+	if t.started {
+		log.DebugF("[%d] Reverse tunnel already up\n", oldId)
+		return -1, fmt.Errorf("already up")
 	}
 
+	id := rand.Int()
```

Comment:
"math/rand" is deprecated and can produce runs of the same values, "crypto/rand" or "math/rand/v2" will be a better choice here

Reply:
done